feat(ingest): Ingestão de Plano de Trabalho Especial (TransfereGov)

marcusmartinss opened the pull request:
### Descrição

Implementação do fluxo de ingestão para o `endpoint /plano_trabalho_especial` da API TransfereGov. O objetivo é capturar os detalhes dos planos de trabalho vinculados aos Planos de Ação já mapeados no sistema.

### Alterações Principais

**1. Plugin (`cliente_transferegov_emendas.py`)**
- Adicionada funcionalidade para a consulta ao endpoint `plano_trabalho_especial`.
- Implementados métodos de paginação manual e automática (`get_all_...`) para garantir a extração completa dos dados.

**2. Nova DAG (`ingest_transferegov_plano_trabalho_especial`)**
- Criação de uma nova DAG que orquestra a ingestão diária.
- Utilização de **Dynamic Task Mapping** para paralelizar as requisições à API baseadas nos IDs de planos de ação.
- Camada Raw: Os dados são salvos em formato JSON achatado (flattened) no Data Lake.

closes #42

marcusmartinss committed:
feat(plugins): adiciona busca de planos de trabalho especial ao cliente TransfereGov

Implementa métodos para buscar planos de trabalho especiais por ID de plano de ação, incluindo suporte a paginação manual e automática.
- get_planos_trabalho_especial_by_plano_acao: busca paginada
- get_all_planos_trabalho_especial_by_plano_acao: busca completa (loop)

marcusmartinss committed:
feat(dags): cria DAG de ingestão para planos de trabalho especial

Adiciona nova DAG 'ingest_transferegov_plano_trabalho_especial' que:
1. Recupera IDs de planos de ação do Postgres.
2. Busca dados na API TransfereGov usando o cliente atualizado.
3. Salva os resultados consolidados na camada Raw.

davi-aguiar-vieira reviewed (changes requested):
A implementação precisa de alguns ajustes para seguir o padrão das DAGs de ingestão do TransfereGov. Seguem os pontos necessários:

---

### **1. Persistência em banco — substituir escrita em arquivo**

A DAG está salvando os dados em JSON na RAW (`save_to_raw`), mas para esse fluxo seguimos o padrão de **inserção direta no Postgres** usando:

```python
db.insert_data(...)
```

Conforme no exemplo `empenhos_especiais_ingest_dag.py`.

Ajustar para:

* Criar um cliente Postgres (`ClientPostgresDB`)
* Adicionar `dt_ingest`
* Inserir na tabela correta (`schema.tabela`)
* Usar `conflict_fields` e `primary_key`

---

### **2. Buscar todos os IDs — remover LIMIT 100**

Na query atual:

```sql
SELECT id_plano_acao FROM transferegov.plano_acao LIMIT 100
```

isso limita os resultados e faz com que parte dos planos de ação não seja processada.

 Favor remover o `LIMIT 100` para coletar **todos** os IDs disponíveis, assim como no exemplo dos empenhos.

---

### **3. Remover estratégia de fallback de IDs**

Hoje, se o Postgres falha, a DAG usa IDs simulados:

```python
return [9060, 9061, 1234]
```

Isso não deve ser utilizado no código final. A DAG deve falhar caso não consiga consultar o banco.

---

### **4. Ajustes menores**

* Corrigir alerta do Sonar sobre concatenação implícita nos logs.
* Manter padrão de logs e docstrings dos outros métodos.
* Validar nomes de schema/tabela no banco.

---

### **5. Ajuste específico**

* Ler o id plano ação do schema `transferegov_emendas`

---

### **6. Conflitos**

* Resolver os conflitos com a main em airflow_lappis/plugins/cliente_transferegov_emendas.py (provavelmente apenas aceitar both changes)

marcusmartinss committed:
Merge remote-tracking branch 'origin/main' into feat/ingestao-do-plano-de-trabalho-especial-transferegov

davi-aguiar-vieira reviewed (changes requested):
### **1. DAG na pasta errada**
A DAG deve estar na pasta [transferegov_emendas](https://github.com/GovHub-br/data-application-gov-hub/tree/feat/ingestao-do-plano-de-trabalho-especial-transferegov/airflow_lappis/dags/data_ingest/transferegov_emendas)

---

### **2. Schema incorreto na inserção**

No `insert_data`, o schema definido é:

```python
schema="transferegov"
```

Mas os IDs estão sendo buscados de:

```sql
transferegov_emendas.plano_acao
```

Por padrão, as ingestões do TransfereGov utilizam o schema:

```
transferegov_emendas
```

Favor confirmar o schema correto e ajustar a linha do insert.

---

### **3. Corrigir alerta do SonarCloud**

O Sonar apontou falha no *Quality Gate* e há issues associadas ao PR.
É necessário revisar e corrigir os pontos sinalizados.

Favor abrir o relatório do SonarCloud no link do PR e resolver todos os problemas em **New Code**, para que o *Quality Gate* volte a passar.



<img width="1918" height="957" alt="image" src="https://github.com/user-attachments/assets/8e219a93-b107-4d52-98df-cd8b4df0b650" />

marcusmartinss committed:
fix(ingest): corrige pipeline TransfereGov e padroniza tabelas

Corrige falhas na ingestão em cascata (Programas -> Ação -> Trabalho) e resolve problemas de conexão e configuração das DAGs.

Erros corrigidos e Mudanças:
- Fix de Dependências: Padronização dos nomes das tabelas para singular ('programa_especial', 'plano_acao') no schema 'transferegov_emendas', resolvendo erros de "relation does not exist".
- Fix de Configuração: Remoção da função 'get_dynamic_schedule' (que causava erro de Broken DAG por falta de config) e fixação do schedule como '@daily'.
- Refactor: Movimentação da DAG de 'plano_trabalho' para a pasta correta ('transferegov_emendas').
- Helper Postgres: Atualização em 'get_postgres_conn' para aceitar 'conn_id' dinâmico e recuperar senhas corretamente da conexão Airflow.
- Plugin Postgres: Implementação do método 'get_all_ids' para facilitar a busca de chaves estrangeiras para tasks mapeadas.

marcusmartinss committed:
fix(sonar): corrige inversão de argumentos em insert_data

Corrige erro de tipagem apontado pelo SonarCloud. A função insert_data espera 'data' como primeiro argumento e 'table_name' como segundo.
A chamada estava invertida. Alterado para argumentos nomeados para garantir a integridade e legibilidade.

davi-aguiar-vieira reviewed (changes requested):
### **1. Remover mudanças em arquivos não relacionados à DAG**

Foram feitas alterações em DAGs que **não fazem parte da issue**, como:

* [programas_especiais_ingest_dag.py](https://github.com/GovHub-br/data-application-gov-hub/blob/main/airflow_lappis/dags/data_ingest/transferegov_emendas/programas_especiais_ingest_dag.py)
* [planos_acao_especiais_ingest_dag.py](https://github.com/GovHub-br/data-application-gov-hub/blob/main/airflow_lappis/dags/data_ingest/transferegov_emendas/planos_acao_especiais_ingest_dag.py)

Essas DAGs já estão em produção há um tempo e **não podemos alterar os nomes das tabelas como feito** ou estruturas existentes, pois:

* Outras DAGs dependem desses nomes atuais
* Mudanças podem quebrar pipelines que já estão em operação

 **Favor desfazer todas as alterações feitas nessas DAGs**, pois não fazem parte da demanda e podem causar impacto indesejado.

---

### **2. A sua DAG de plano de trabalho deve consumir IDs da tabela correta**

A DAG de plano de trabalho especial deve buscar os IDs de **plano de ação especial** a partir da tabela:

```
transferegov_emendas.planos_acao_especiais
```

Atualmente está consultando outra tabela — favor corrigir a origem dos IDs conforme o modelo vigente.

---

### **3. Remover alterações no helper do Postgres**

Houve mudanças no arquivo:

- [postgres_helpers.py](https://github.com/GovHub-br/data-application-gov-hub/blob/main/airflow_lappis/helpers/postgres_helpers.py)

Essas alterações **não fazem parte do escopo da issue**, não são necessárias para tal e podem impactar outras partes do projeto.

**Favor reverter as modificações nesse helper.**

---

### **4. Resolver conflito com a base branch (main)**

O PR está apresentando conflitos com a `main`.
Por favor, faça o merge/rebase da branch e resolva os conflitos antes da próxima revisão.

---

Assim que esses ajustes forem feitos, estou a disposição para revisar novamente. Obrigado!
